annotation-server/src/guidelines/guidelines.service.ts

```diff
@@ -0,0 +1,322 @@
+            return this.hashedMedicationsByRxCUI.get(rxcui);
+        }
+        try {
+            const medication = await this.medicationsService.getOne({
+                where: {
+                    rxcui,
+                },
+            });
+            this.hashedMedicationsByRxCUI.set(rxcui, medication);
+            this.hashedMedicationsByName.set(medication.name, medication);
+            return medication;
+        } catch (error) {
+            this.logger.error(
+                `Medication with RxCUI ${rxcui} not found in our database.`,
+            );
+            return null;
+        }
+    }
+
+    private async findGenePhenotype(
+        geneSymbolName: string,
+        lookupkey: string,
+    ): Promise<GenePhenotype> {
+        if (!geneSymbolName || !lookupkey) return null;
+        geneSymbolName = geneSymbolName.trim().toLowerCase();
+        const genePhenotype =
+            await this.genePhenotypesService.getOneGenePhenotype({
+                where: {
+                    geneSymbol: { name: ILike(geneSymbolName) },
+                    phenotype: { lookupkey },
+                },
+                relations: ['phenotype', 'geneSymbol'],
+            });
+        return genePhenotype;
+    }
+
+    private async findGenePhenotypes(
+        geneSymbolName: string,
+    ): Promise<Array<Set<GenePhenotype>>> {
+        if (!geneSymbolName) return [];
+        geneSymbolName = geneSymbolName.trim().toLowerCase();
+        if (this.hashedGenePhenotypes.has(geneSymbolName)) {
+            return this.hashedGenePhenotypes.get(geneSymbolName);
+        }
+        const geneSymbol = await this.genePhenotypesService.getOne({
+            where: { name: ILike(geneSymbolName) },
+            relations: ['genePhenotypes', 'genePhenotypes.phenotype'],
+        });
+        // TODO: consider proper error handling
+        if (!geneSymbol) {
+            this.logger.error(
+                `Gene ${geneSymbolName.toUpperCase()} not found in CPIC lookupkeys.`,
+            );
+            this.hashedGenePhenotypes.set(geneSymbolName, []);
+            return [];
+        }
+        const genePhenotypes = this.spreadsheetPhenotypeHeader.map(
+            (phenotypes) =>
+                new Set(
+                    geneSymbol.genePhenotypes.filter((genePhenotype) =>
+                        phenotypes.has(
+                            genePhenotype.phenotype.name.toLowerCase(),
+                        ),
+                    ),
+                ),
+        );
+        this.hashedGenePhenotypes.set(geneSymbolName, genePhenotypes);
+        return genePhenotypes;
+    }
+
+    private getIncompleteGuidelines(guidelines: Guideline[]): Guideline[] {
+        const incompleteGuidelines: Guideline[] = [];
+        for (const guideline of guidelines) {
+            if (!guideline.implication && !guideline.recommendation)
+                incompleteGuidelines.push(guideline);
+        }
+        return incompleteGuidelines;
+    }
```

Comment:
I think it might be cool to add this as a method (might have to be static because of TypeORM) on the `Guideline` entity, e.g. `isComplete: bool`. If that goes against TypeORM conventions I would keep it here but change it to a simple `filter` one-liner for better optimization.

Reply:
I feel like the entity should be relatively small and without a lot of knowledge, so I would intuitively prefer keeping the method in the service. 
In the [TypeORM docs](https://typeorm.io/entities) it only says "Entity is a class that maps to a database table", which could be interpreted as "all fields/methods have to map to fields in the database", but that is not stated explicitly.

Reply:
I tried doing some research but couldn't find any bias or reasoning on whether or not TypeORM Entities can / should have methods. I did however find [this StackOverflow answer](https://stackoverflow.com/a/62053220) that shows Entities can have getters, i.e. computed properties. From a mere OO-perspective I would prefer a computed `isComplete` property on the Entity by a long shot. All the data required to determine a Guideline Entity's completeness is stored in the Guideline Entity itself which is why it doesn't make sense to me to have it anywhere else (separation of concerns). This also is analogous to how we have defined static constructors on the Guideline and Medication Entities.